pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+	if !inFlight {
+		r.addWorkItems(rev, opt, registriesToSkip, timeout)
+		return nil, nil
+	}
+
+	if !result.ready() {
+		return nil, nil
+	}
+
+	return r.results[name].statuses, r.results[name].err
+}
+
+// addWorkItems adds a digest resolve item to the queue for each container in the revision.
+// This is expected to be called with the mutex locked.
+func (r *backgroundResolver) addWorkItems(rev *v1.Revision, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) {
+	name := types.NamespacedName{
+		Name:      rev.Name,
+		Namespace: rev.Namespace,
+	}
+
+	r.results[name] = &resolveResult{
+		statuses:  make([]v1.ContainerStatus, len(rev.Spec.Containers)),
+		remaining: len(rev.Spec.Containers),
+		completionCallback: func() {
+			r.enqueue(name)
+		},
+	}
+
+	for i := range rev.Spec.Containers {
+		image := rev.Spec.Containers[i].Image
+
+		r.queue.Add(&workItem{
+			result:  r.results[name],
+			timeout: timeout,
+			name:    rev.Spec.Containers[i].Name,
+			image:   image,
+			index:   i,
+		})
+	}
+}
+
+// processWorkItem runs a single image digest resolution and stores the result
+// in the resolveResult. If this completes the work for the revision, the
+// completionCallback is called.
+func (r *backgroundResolver) processWorkItem(item *workItem) {
+	ctx, cancel := context.WithTimeout(context.Background(), item.timeout)
+	defer cancel()
+
+	result, err := r.resolver.Resolve(ctx, item.image, item.result.opt, item.result.registriesToSkip)
+
+	// lock after the resolve because we don't want to block parallel resolves,
+	// just storing the result.
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// If we're already ready we don't want to callback twice.
+	// This can happen if an image resolve completes but we've already reported
+	// an error from another image in the result.
+	if item.result.ready() {
+		return
+	}
+
+	if err != nil {
+		item.result.statuses = nil
+		item.result.err = ContainerMissingError{image: item.image, cause: err}
+	} else {
+		item.result.remaining--
+		item.result.statuses[item.index] = v1.ContainerStatus{
+			Name:        item.name,
+			ImageDigest: result,
+		}
+	}
+
+	if item.result.ready() {
+		item.result.completionCallback()
+	}
+}
+
+// Clear removes any cached results for the revision. This should be called
+// when the revision is deleted or once the revision's ContainerStatus has been
+// set.
+func (r *backgroundResolver) Clear(name types.NamespacedName) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// delete the map entry, the result can still be accessed via the pointer in
+	// the workItem so processWorkItem doesn't care about this, but once the
+	// workItem is done the result will be GC-ed.
+	delete(r.results, name)
+}
+
+func (r *resolveResult) ready() bool {
+	return r.remaining == 0 || r.err != nil
+}
+
+// ContainerMissingError converts an error in to the expected format for the
+// RevisionContainerMissing condition.
+type ContainerMissingError struct {
```

Comment:
is there a reason to have this type public? In the end we just return `error`?

Reply:
not really, I quite like errors being public so you can do `errors.Is` on them, but happy to make it private

Reply:
You can do `Is` anyways on this as you implement unwrap, but I don't think this type being public gives you anything in that regard, unless you want to type-cast it somewhere.

Reply:
I mean it needs to be pubic so you can do `errors.Is(revision.ContainerMissingError, err)`, if it's not public you can't do that outside the package

Reply:
(it's just otherwise people are tempted to start matching on the error string)

Reply:
Can you really? You can do that with error constants, like `const ErrContainerMissing = errors.New("bla")`. But does that work with just types too? :thinking:

Reply:
from slack convo: you can do `errors.As` but also let's just make it private :D